GitUI/UserControls/RevisionGrid/FilterInfo.cs

```diff
@@ -423,32 +423,29 @@ private void GetBranchRevisionFilter(ArgumentBuilder filter)
                 // All commits
                 filter.Add("--reflog");
             }
-            else if (IsShowCurrentBranchOnlyChecked || (IsShowFilteredBranchesChecked && !string.IsNullOrWhiteSpace(BranchFilter)))
+            else if (IsShowCurrentBranchOnlyChecked && !string.IsNullOrWhiteSpace(currentBranch.Value))
             {
-                // See handling for --all below, explicitly add most recent stash
+                // Git default, the branch must exist.
+                // No option added except for stashes.
+
                 if (AppSettings.ShowStashes)
                 {
-                    // stash@{0} requires that the repo has commits and --glob=refs/stash is handled as refs/stash/*,
-                    // so this weird pattern must be used: "--glob=refs/stas[h]"
-                    filter.Add($"--glob={GitRefName.RefsStashPrefix.Remove(GitRefName.RefsStashPrefix.Length - 1, 1)}[h]");
-                }
+                    AddFirstStashRef();
 
-                if (IsShowCurrentBranchOnlyChecked)
-                {
-                    // Default git-log (no option), only current branch (HEAD)
-                    if (AppSettings.ShowStashes)
-                    {
-                        filter.Add("HEAD");
-                    }
-                }
-                else
-                {
-                    // Show filtered branches only
-                    // Do not add suffix /* to the filter
-                    bool isSimpleBranchFilter = BranchFilter.IndexOfAny(new[] { '?', '*', '[' }) == -1;
-                    filter.Add(isSimpleBranchFilter ? BranchFilter : $"--branches={BranchFilter}");
+                    // current branch must be an option due to --glob
+                    filter.Add(currentBranch.Value);
                 }
             }
+            else if (IsShowFilteredBranchesChecked && !string.IsNullOrWhiteSpace(BranchFilter))
+            {
+                // Show filtered branches only
+                AddFirstStashRef();
+
+                // Simple branch filter without wildcards
+                // Git adds the suffix "/*" implicitly if there is no wildcard
+                bool isSimpleBranchFilter = BranchFilter.IndexOfAny(new[] { '?', '*', '[' }) == -1;
+                filter.Add(isSimpleBranchFilter ? BranchFilter : $"--branches={BranchFilter}");
```

Comment:
Should we really keep this check for simple filter? Has git changed its behavior? Could be changed here or in a follow-up.

![image](https://user-images.githubusercontent.com/36601201/225153868-7dc39cf7-0fb3-47c9-a2fa-2710e813f67e.png)

Reply:
https://git-scm.com/docs/git-log#Documentation/git-log.txt---branchesltpatterngt

I assume this is a branch filter.
This PR is only changing the behavior for currentbranchonly.
The same popup, but the branch names should be validated already in the form (toolbar or advanced, normally testing the cached gitrefs first) and give a feedback that the selection is invalid.
That is a separate PR.

Reply:
> This PR is only changing the behavior for currentbranchonly.

As stated:

>> Could be changed ... in a follow-up.

My point is: The comment `'Git adds the suffix "/*" implicitly if there is no wildcard'` seems to be wrong or does not apply to how the arguments are added here.

Reply:
> My point is: The comment 'Git adds the suffix "/*" implicitly if there is no wildcard' seems to be wrong or does not apply to how the arguments are added here.

The wildcard is added with --branches, not as a ref on the command line. 
Clarification?

// Git adds the suffix "/*" implicitly if there is no wildcard for --branches argument

The same 'fix' that is used for refs/stas[h] could be used here too. Then the issue with invalid references for branches could be solved relatively simple. The drawback is that it is not possible to apply any ref (or option) in the branchbox, including multiple branches (multi could be split).
It is still possible to use messages with --init --exclude prefix (_messageAsGitOptions), but that is a special feature...

--branches=<currentbranch> is an alternative for currentbranchonly, no validation of the branch is needed (so solution is more similar to the first version).

Reply:
Since we use `--branches=` only if the branch filter already contains wildcards, I do not see the relevance of the comment - although the proposed variant is clearer.
Or does it mean that we want to avoid adding `--branches=` in order to not get the implicit "/*" appended by git?
(I don't have a preference whether branch filters should be an exact match or not. I just want to understand the intention of the code.)

Reply:
I mean that `feature` could be added as `--branches=featur[e]` similar to refs/stash

Reply:
> I mean that `feature` could be added as `--branches=featur[e]` similar to refs/stash

This stops git from returning errors. 👍
But it doesn't work for commit hashes and HEAD, etc. 🤔

Reply:
There were actually an error when a repo was empty and stashes was shown, so I changed to always use --branches for currentbranchonly.
Actually slightly simpler code.
(I thought I had tested that, I have updated Git since then, I doubt it is that though...)

To be clear, it is the Git error code we must avoid, the behavior in the app is the same. (Just to avoid the popup, no other improvement for the user.)

Filtered branches is a separate issue. It could be possible to use similar command as here, but maybe generic refs like commits should be allowed?

Reply:
@mstv Still OK with the latest changes?
If so I will merge this and look into filtered branches (probably evaluating when exiting the text box using the ref filter to verify before rev-parse)